### PR TITLE
[1/3]: Preparatory work for Forwarding Blinded Routes

### PR DESCRIPTION
**Pull request by carlaKC**
This PR is the first in a series to add the ability for LND to forward payments to blinded routes.

It contains a collection of prep work to get the codebase into shape for forwarding functionality: 
* Add encoding/decoding and validation for `encrypted_data` that is provided for blinded forwards. 
* Add `blinding_point` to `update_add_htlc` and persist across restarts.
* Small refactors to prepare for decoding onion payloads with blinding points. 

**Note**: the ability to generate/receive payments via blinded routes is out of scope for these PRs.

### CODE REVIEW

File: lnwire/features.go [hunk 1]
```diff
@@ -612,6 +614,41 @@ func (fv *RawFeatureVector) decode(r io.Reader, length, width int) error {
...
 
+// sizeFunc returns the length required to encode the feature vector.
+func (fv *RawFeatureVector) sizeFunc() uint64 {
+	return uint64(fv.SerializeSize())
+}
+
+// Record returns a TLV record that can be used to encode/decode raw feature
+// vectors. Note that the length of the feature vector is not included, because
+// it is covered by the TLV record's length field.
+func (fv *RawFeatureVector) Record(recordType tlv.Type) tlv.Record {
```
Comment (ellemouton) [inline lnwire/features.go:625]: 👍

File: record/blinded_data.go [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+const (
+	// ShortChannelIDType is a record type for the outgoing channel short
+	// ID.
+	ShortChannelIDType tlv.Type = 2
```
Comment (ellemouton) [inline record/blinded_data.go:16]: with the new `tlv.RecordT` and `tlv.OptionalT` types, you can make the tlv number part of the field type itself like: 

```
ShortChannelID tlv.RecordT[tlv.Type2, lnwire.ShortChannelID]
```

See examples [here](https://github.com/lightningnetwork/lnd/pull/8338) and [here](https://github.com/lightningnetwork/lnd/pull/8044)

File: htlcswitch/hop/payload.go [hunk 3]
```diff
@@ -380,3 +387,71 @@ func getMinRequiredViolation(set tlv.TypeMap) *tlv.Type {
...
+
+// ValidateBlindedRouteData performs the additional validation that is
+// required for payments that rely on data provided in an encrypted blob to
+// be forwarded. We enforce additional constraints here to prevent malicious
+// parties from probing portions of the blinded route to "un-blind" them.
+func ValidateBlindedRouteData(blindedData *record.BlindedRouteData,
+	incomingAmount lnwire.MilliSatoshi, incomingTimelock uint32) error {
+
+	// Bolt 04 notes that we should enforce payment constraints _if_ they
+	// are present, so we don not fail if not provided.
```
Comment (ellemouton) [inline htlcswitch/hop/payload.go:399]: nit: s/don/do

File: lnwire/update_add_htlc.go [hunk 4]
```diff
@@ -54,6 +54,10 @@ type UpdateAddHTLC struct {
 	// used in the subsequent UpdateAddHTLC message.
 	OnionBlob [OnionPacketSize]byte
 
+	// BlindingPoint is the ephemeral pubkey used to optionally blind the
+	// next hop for this htlc.
+	BlindingPoint *blindingPoint
```
Comment (ellemouton) [inline lnwire/update_add_htlc.go:59]: can use `tlv.OptionalRecordT[tlv.TlvType0, blindingPoint]`

File: lnwallet/channel.go [hunk 5]
```diff
@@ -879,6 +894,12 @@ func (lc *LightningChannel) diskHtlcToPayDesc(feeRate chainfee.SatPerKWeight,
 		theirWitnessScript: theirWitnessScript,
 	}
 
+	var err error
+	pd.BlindingPoint, err = htlc.BlindingPoint()
+	if err != nil {
+		return pd, nil
```
Comment (ellemouton) [inline lnwallet/channel.go:900]: return pd, err?

**Review by ellemouton — COMMENTED** [review]
Very nice & clean! First pass done. Basically LGTM . Main comment is re potentially using the new `tlv.RecordT` and `tlv.OptionalRecordT` types for the various new tlv fields.

File: lnwallet/channel.go (hunk 5 above)
Comment (carlaKC) [reply lnwallet/channel.go:900]: oof, rebase booger. Nice catch!

**Comment by carlaKC** [comment]
Just pushed rebase + small fixups, looking into moving TLVs over to generics!

**Comment by ellemouton** [comment]
Just following up on my previous comment: 

I personally think it's fine for us to move ahead with the current structure of the new TLV fields. Better to stick to the current "standard" way and then leave the big re-hall of _all_ TLV fields to another PR. Else we will need to reason about 2 different ways of doing the same thing. 

Let me know if you agree with this @carlaKC  - then i'll re-request myself here for final review which I will prio asap 🚀

**Comment by carlaKC** [comment]
> I personally think it's fine for us to move ahead with the current structure of the new TLV fields. Better to stick to the current "standard" way and then leave the big re-hall of _all_ TLV fields to another PR. Else we will need to reason about 2 different ways of doing the same thing.

Started working on switching over to generics [here](https://github.com/lightningnetwork/lnd/commit/e26a3dd3bae098f6e6e7292fa993b3dde6aeb038), and I agree that It may not be worth trying to make this jump now. From what I can tell generics work isn't _totally_ there, so it's a bit patchwork adding it in this PR IMO.

> Let me know if you agree with this @carlaKC - then i'll re-request myself here for final review which I will prio asap 🚀

I'll rebase then re-request you, thanks!!

**Review by ellemouton — COMMENTED** [review]
Really great stuff 🚀 🧅 

Main question is around if we need to account for `next_blinding_override` in order to be spec compatible. 

Will come back for final look at this PR once i've left first review on the next one as I just want to verify for myself if all checks i've noted from the spec are in place 👍 Main one on my mind is: do we check that the payment relay constraints set by the recipient actually match ones that we have recently advertised? or is that just implicitly checked elsewhere in the normal relay flow? 

(btw, Not sure if @Roasbeef's comments on the [temp generics commit](https://github.com/lightningnetwork/lnd/commit/e26a3dd3bae098f6e6e7292fa993b3dde6aeb038#r138743808) would have generated a notification so just gonna link that here too)

**Comment by carlaKC** [comment]
> do we check that the payment relay constraints set by the recipient actually match ones that we have recently advertised? or is that just implicitly checked elsewhere in the normal relay flow?

Indeed, once we've magic-ed our forwarding values in they'll be checked in the normal flow of things. 

> (btw, Not sure if @Roasbeef's comments on the [temp generics commit](https://github.com/lightningnetwork/lnd/commit/e26a3dd3bae098f6e6e7292fa993b3dde6aeb038#r138743808) would have generated a notification so just gonna link that here too)

Will check this out, shouldn't be a material change to the logic of the PR if we switch to generics.

File: htlcswitch/hop/payload.go [hunk 6]
```diff
@@ -410,3 +417,71 @@ func getMinRequiredViolation(set tlv.TypeMap) *tlv.Type {
...
+	incomingAmount lnwire.MilliSatoshi, incomingTimelock uint32) error {
+
+	// Bolt 04 notes that we should enforce payment constraints _if_ they
+	// are present, so we do not fail if not provided.
+	if blindedData.Constraints != nil {
+		// MUST fail if the expiry is greater than max_cltv_expiry.
+		//
+		// TODO: confirm handling of zero values in spec, at the time
+		// of writing some implementations are providing zero values
+		// to mean "no restriction".
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply htlcswitch/hop/payload.go:437]: Going to check up on this again - ran into it while interop testing but that was a while ago (could have been a bug in somebody else's impl)

**Comment by carlaKC** [comment]
Addressed feedback (incl `next_blinding_override`), ready for another look! 

Things that I owe on this: 
1. [Clarification](https://github.com/lightningnetwork/lnd/pull/8159#discussion_r1492392587) of use of zero values in the spec. 
2. Proper investigation into generics.

**Comment by carlaKC** [comment]
> https://github.com/lightningnetwork/lnd/pull/8159#discussion_r1492392587 of use of zero values in the spec.

Confirmed in spec meeting that we shouldn't have zero values floating around, so I'll pull that out and re-test interop (possible that the zero value was a bug in CLN which has since been fixed).

**Review by ellemouton — COMMENTED** [review]
Looks good! left a few questions for my understanding. 

Think this is ready for green tick but sounds like (from your last comment) that you are planning to still push the generics change so will wait for that before final ACK ✅

**Review by bitromortac — COMMENTED** [review]
Looks awesome on first pass :fire:, I will need to familiarize more with the changes though.

File: channeldb/channel.go [hunk 7]
```diff
@@ -2224,6 +2224,58 @@ type HTLC struct {
 	ExtraData []byte
 }
 
+// SetExtraData encodes a TLV stream of extra data to be stored with a HTLC.
+// It uses the update_add_htlc TLV types, because this is where extra data is
+// passed with a HTLC. At present blinding points are the only extra data that
+// we will store, and the function is a no-op if a nil blinding point is
+// provided.
+func (h *HTLC) SetExtraData(blindingPoint *btcec.PublicKey) error {
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply channeldb/channel.go:2232]: > I assume the reason is to have the HTLC struct closely represent the schema used on disk.

Yeah, reasoning was to only have the fields that are persisted on the `channeldb.HTLC` struct, since we're already using a kinda odd workaround to store the data. 

> instead of just *BlindedPoint and then just handling the persistence details in the HTLC encoding/decoding methods

Looking at this again, this seems to be the approach that we take for messages with TLVs like `AcceptChannel`, so would be happy to update to match that? Could also pair it with un-exporting `ExtraData` on `channeldb.HTLC` which could be a little cleaner 🤔

**Comment by carlaKC** [comment]
Pushed the changes suggested, thanks for the review! Left the fixups _unsquashed_ but next to their original commit so that you can view the delta (GH lets you select the two commits together to view the full diff) so the changes are clear (shout if that's useful/annoying). 

One last thing that I think we may want to change is [this comment](https://github.com/lightningnetwork/lnd/pull/8159#discussion_r1505523986) from @ellemouton, I think that leaving details in encode/decode may make more sense now that I've looked at the rest of the codebase. LMK if we should switch and I'll get on it!

> Re: Generics

I had best intentions to get this PR switched over, but (shame on me) given the RC deadline I think it's best to just leave it to a followup.

**Review by ellemouton — APPROVED** [review]
Great work! 🚗 🔍 

Makes sense re follow up for generic given RC time pressure :)

File: htlcswitch/hop/payload.go [hunk 8]
```diff
@@ -410,3 +417,85 @@ func getMinRequiredViolation(set tlv.TypeMap) *tlv.Type {
...
+		return ErrInvalidPayload{
+			Type:      record.PaymentRelayType,
+			Violation: OmittedViolation,
+		}
+	}
+
+	// Specification validation instructions for receivers are:
+	// * encrypted data must either have short_channel id _or_ next_node_id.
+	//
+	// However, instructions for creators of blinded route encrypted data
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply htlcswitch/hop/payload.go:462]: > my main question is why does the spec allow for both and not just explicitly disallow it

Could be an oversight, will double check!

> So yeah - i cant think of a reason why we would ever want or allow the receiver to set node id in future.

Same, I think it's just for onion messaging - don't see us adding it for payments.

**Comment by Roasbeef** [comment]
> I had best intentions to get this PR switched over, but (shame on me) given the RC deadline I think it's best to just leave it to a followup.

What significant barriers do you see? It's just using a more declarative way of specifying TLV fields, and in most cases would result in less code, and also more explicit `nil` checks, which can save us from DoS issues down the line. If you're looking for an inspo PR, you can check out this PR where we recently converted the nonce TLVs for taproot to use the new APIs: https://github.com/lightningnetwork/lnd/pull/8499. 

Given the third PR is still in draft, I have a hard time believing that the API switch would delay final inclusion vs new interactions or considerations uncovered ruing review of the final PR.

**Comment by carlaKC** [comment]
> What significant barriers do you see? It's just using a more declarative way of specifying TLV fields

As when I took a first look at using the new generics code, I feel like we're at risk of needing to come back and refactor _anway_ if it changes so rather just leave it for the time when we switch over the codebase more generally.

I've gone through the full PR series to [upgrade to generics ](https://github.com/carlaKC/lnd/tree/7298-generics) and ran into a few sticky points (changes in unsquashed fixups). Perhaps it's my own lack of understanding, or rust mental model overlapping but I struggled with the following:
* [ExtraData oriented](https://github.com/carlaKC/lnd/commit/8e57a079686de363edf8ff1b897e70d61aaeed7c#diff-517e67d2397e5aca9d2af30be30b98138190a4c0211fe619ac2d0e42ef7ee622R123): seems like we have so far only used the generics in extension TLVs (not native TLV structs like) so there aren't regular `tlv.Stream` utilities? Everything needs to be converted to `ExtraOpaqueData` and then packed. Not a lot of code, but is this how we want to handle things?
* [Non-Primitive types](https://github.com/carlaKC/lnd/commit/8e57a079686de363edf8ff1b897e70d61aaeed7c#diff-517e67d2397e5aca9d2af30be30b98138190a4c0211fe619ac2d0e42ef7ee622R22): while we have `RecordT` for primitive types, there don't seem to be the same for non-primitive TLVs? Hit this with feature vectors in this PR: we can't implement `Record` because features are encoded with different tlv types in different contexts, and there isn't a simple way to use generics _without_ having  implemented `RecordProducer`, so we still need boilerplate. 
* Adapting `Option`: handling optional records is quite clumsy? In rust, you can just `if let Some(A)` to access the inner value  _when it is present_ and remain in the scope of your current function. With the existing `Unwrap` utilities we're always using closures to do simple things [like validate`Some(A)`](https://github.com/carlaKC/lnd/commit/b65fab88fd2bf0c1f742211cc424cf014ab3d199#diff-0f9dab9f73a179ef469f38afe4c4da569632a0b93d75be4a09024311da62ed64R433) _if_ it's non-nil.

That said, I've updated the whole series now so I can happily squash that branch and push it for review. I'd really like to get as much of this in as possible for the RC, so happy to take whatever path gets us there. 

> Given the third PR is still in draft, I have a hard time believing that the API switch would delay final inclusion vs new interactions or considerations uncovered ruing review of the final PR.

Based on previous conversations with @saubyk in contributor calls, we'd spoken about the possibility merging 1+2 and not 3 (so that we don't signal the feature) if time gets tight, so the slowdown would be extra rounds of review on the first two.

**Review by bitromortac — COMMENTED** [review]
Looks great :tada:! I think we should just double check that nobody relies on pubkey-based forwarding (seems unlikely due to the receiver specs, but can't hurt). I'm also happy to re-review if you decide to include the TLV-related changes.

**Comment by carlaKC** [comment]
> I think we should just double check that nobody relies on pubkey-based forwarding (seems unlikely due to the receiver specs

Confirmed that nobody's using pubkey-based forwarding, and that we can tighten up the specification to only allow a `short_channel_id` or `next_node_id`. Since recipient paths are only valid with a `short_channel_id`, I don't think that we need to have a forwarding check as well.

**Comment by ellemouton** [comment]
>  we can't implement Record because features are encoded with different tlv types in different contexts, and there isn't a simple way to use generics without having implemented RecordProducer, so we still need boilerplate.

Left a comment on your generics commit about this: https://github.com/carlaKC/lnd/commit/e8b47d431e111c86272692225dff782ec6147608?diff=split&w=0#r139714467

I think it is possible to do this without repeating the boilerplate. See [this](https://github.com/lightningnetwork/lnd/pull/8044/commits/ae4b427caf8b024f7934c9c97216f7a343346182) commit for an example

**Comment by carlaKC** [comment]
> Left a comment on your generics commit about this: [carlaKC@e8b47d4?diff=split&w=0#r139714467](https://github.com/carlaKC/lnd/commit/e8b47d431e111c86272692225dff782ec6147608?diff=split&w=0#r139714467)

Ah, nice! Awaiting the bat signal re whether to switch the series over to generics or not, I'll add this + do a cleanup runthrough if we're going that route 👍

**Comment by carlaKC** [comment]
Just a rebase!

**Review by calvinrzachman — APPROVED** [review]
IT'S HAPPENING 🙌 Just a couple questions on validation.

File: htlcswitch/hop/payload.go [hunk 9]
```diff
@@ -410,3 +417,85 @@ func getMinRequiredViolation(set tlv.TypeMap) *tlv.Type {
...
+	// No need to check anything else if features are not provided (bolt 4
+	// indicates that omitted features should be treated like an empty
+	// vector).
+	if blindedData.Features == nil {
+		return nil
+	}
+
+	// Fail if we don't understand any features (even or odd), because we
+	// expect the features to have been set from our announcement.
+	if blindedData.Features.UnknownFeatures() {
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply htlcswitch/hop/payload.go:493]: The creator specs say that if `allowed_features` is set, it must set it to an empty array, but in the forwarding case we are less restrictive, to be able to set some features in the future I guess. The question is then to which features we should compare to, to the node's feature set or to the channel's feature set (there are currently none specified https://github.com/lightning/bolts/blob/3b814f0d031450a9cddd2718be7ed50c5444095b/09-features.md)?

File: htlcswitch/hop/payload.go [hunk 10]
```diff
@@ -410,3 +417,85 @@ func getMinRequiredViolation(set tlv.TypeMap) *tlv.Type {
...
+		// MUST fail if the amount is below htlc_minimum_msat.
+		if incomingAmount < blindedData.Constraints.HtlcMinimumMsat {
+			return ErrInvalidPayload{
+				Type:      record.AmtOnionType,
+				Violation: InsufficientViolation,
+			}
+		}
+	}
+
+	// Encrypted data MUST include payment relay information.
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply htlcswitch/hop/payload.go:451]: > Maybe we're fine to save this validation for when we add the ability to receive!

Yeah, leaving that till we're supporting receives 👍

File: htlcswitch/hop/payload.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (carlaKC) [reply htlcswitch/hop/payload.go:493]: >  to be able to set some features in the future I guess.

Yeah this is future proofing us for upgrades to things like PTLCs, where we want to make sure we support the payment type. Since creators can't set the feature, I think we can leave this out of scope (at least until LND is creating its own blinded paths?)

> one could check if tlv onions are advertised and then not forward if encountered

I'll clarify this is the spec, but I _think_ that `tlv_onion` doesn't need to be enforced here as it's implicitly supported by our ability to parse the payload in the first place?

File: lnwallet/channel_test.go [hunk 11]
```diff
@@ -10802,3 +10803,43 @@ func TestEnforceFeeBuffer(t *testing.T) {
...
+	_, err = bobChannel.ReceiveHTLC(htlc)
+	require.NoError(t, err)
+
+	// Now, Alice will send a new commitment to Bob, which will persist our
+	// pending HTLC to disk.
+	_, err = aliceChannel.SignNextCommitment()
+	require.NoError(t, err, "unable to sign commitment")
+
+	// Restart alice to force fetching state from disk.
+	aliceChannel, err = restartChannel(aliceChannel)
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply lnwallet/channel_test.go:10838]: noice 🚀

**Review by bitromortac — APPROVED** [review]

**Comment by Roasbeef** [comment]
> seems like we have so far only used the generics in extension TLVs (not native TLV structs like) so there aren't regular tlv.Stream utilities? Everything needs to be converted to ExtraOpaqueData and then packed. Not a lot of code, but is this how we want to handle things?

Nope, you can just serialize into a normal blob. Look at this commit where I encoded things into a normal blob: https://github.com/lightningnetwork/lnd/pull/8546/commits/784d236974794f12bfd678c441debb12687dff9c. It's just like before, you make a stream form the records, but now the attributes themselves are records. 

> while we have RecordT for primitive types, there don't seem to be the same for non-primitive TLVs? Hit this with feature vectors in this PR: we can't implement Record because features are encoded with different tlv types in different contexts, and there isn't a simple way to use generics without having implemented RecordProducer, so we still need boilerplate.

You can make it from anything that implements record producer: https://github.com/lightningnetwork/lnd/blob/b0552da0073b1309524a20f18303ce42f69e868f/tlv/record_type.go#L34-L44. So for this, we'd have a single struct that wraps the feature vector to implement the `Record()` method, then that single struct can be re-used across any other attributes. 

> Adapting Option: handling optional records is quite clumsy? In rust, you can just if let Some(A) to access the inner value when it is present and remain in the scope of your current function. With the existing Unwrap utilities we're always using closures to do simple things [like validateSome(A)](https://github.com/carlaKC/lnd/commit/b65fab88fd2bf0c1f742211cc424cf014ab3d199#diff-0f9dab9f73a179ef469f38afe4c4da569632a0b93d75be4a09024311da62ed64R433) if it's non-nil.

It's what we have right now, but doesn't at all appear to be a fundamental barrier. You can either revert back to using plain pointers, then be extra diligent about nil checks, use the existing utilities, or make your own along the way to mske things more egonomic. You can use `WhenSome` for when you want to run a chunk of logic only if the thing is there, the `Map` functions for when you want to have a return value, and the other hard checks if you want to force an error if something is non-nil `UnwrapOrErr`. The latter is pretty much the same amount of lines as a `ptr != nil` check. 

In short, nothing you've mentioned appear to be a significant barrier. Particularly given that many of PRs and newly introduce sub-systems have started to use the new TLV records. I've pointed out areas where we can delete entire files, and also make `nil` checks more explicit, to avoid a dreaded HTLC of death panic error, or other DOS issues. I really don't see any reason for switching over in this PR. In the amount of back and forth, you likely could've just made the routine API switch. Compared to everything else in this PR, it's rather minor, and works to make TLV handling more uniform, and also less error prone (forgot to do a nil check for optional TLV, causing a panic).

**Comment by carlaKC** [comment]
> In the amount of back and forth, you likely could've just made the routine API switch.

While I absolutely could have updated the PR to use generics without full confidence that it was the right decision as the author, I don't really think that's how we want to be writing code. I appreciate the time that you and @ellemouton have taken to point out where I've been wrong about my concerns, and have throughout this discussion made a good faith attempt to maintain a version that uses generics that I feel this comment conveniently handwaves away. 

I still have my reservations about directly porting rust primitives over to golang implementations, but I recognize that I should have reviewed the PRs in question if I wanted a vote. I have switched the series over to use generics - thanks in advance to reviewers for the extra round / apologies for making it necessary.

File: lnwire/features.go [hunk 12]
```diff
@@ -678,6 +678,50 @@ func EmptyFeatureVector() *FeatureVector {
...
 }
 
+// Record implements the RecordProducer interface for FeatureVector. Note that
+// it uses a zero-value type is used to produce the record, as we expect this
+// type value to be overwritten when used in generic TLV record production.
+// This allows a single Record function to serve in the many different contexts
+// in which feature vectors are encoded. This record wraps the encoding/
+// decoding for our raw feature vectors so that we can directly parse fully
+// formed feature vector types.
+func (fv *FeatureVector) Record() tlv.Record {
```
Comment (ellemouton) [inline lnwire/features.go:688]: I think we can get around needing this by converting the `RawFeatureVector` `Record` method to satisfy the RecordProducer interface and then just using `Features tlv.OptionalRecordT[tlv.TlvType14, lnwire.RawFeatureVector]` in the struct.

Comment (ellemouton) [reply lnwire/features.go:688]: Tried it locally & seems to work fine:

[rawFeature.patch](https://github.com/lightningnetwork/lnd/files/14647157/rawFeature.patch)

**Review by ellemouton — APPROVED** [review]
Update to generics LGTM! Just left one suggestion that would reduce the number of lines changed a bit 👍

File: lnwire/features.go (hunk 12 above)
Comment (carlaKC) [reply lnwire/features.go:688]: Yeah we can! 

@bitromortac's [comment](https://github.com/lightningnetwork/lnd/pull/8159#discussion_r1528308207) made me realize that long term it'll be nice to have the full feature vector so we can use some of the utility methods that we have on `FeatureVector` when we get to the stage where we have to check the features in the blinded path vs the payment features. 

With raw feature vector, we parse it out and then have to call `NewFeatureVector` afterwards to use things like `HasFeature`. But no strong opinions here tbh!

Comment (ellemouton) [reply lnwire/features.go:688]: oh cool good point! gotcha 👍 

yeah i dont have strong opinions here either so happy for it to be left as is for those future checks 🎉

File: channeldb/channel.go [hunk 13]
```diff
@@ -35,6 +35,10 @@ const (
 	// begins to be interpreted as an absolute block height, rather than a
 	// relative one.
 	AbsoluteThawHeightThreshold uint32 = 500000
+
+	// HTLCBlindingPointTLV is the tlv type used for storing blinding
+	// points with HTLCs.
+	HTLCBlindingPointTLV tlv.Type = 0
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply channeldb/channel.go]: Leftover from pre-generics!

>  I think that would make sense to keep it decoupled from the wire message tlv type

Also originally thought this, but I don't really see us adding anything in here that _isn't_ from `update_add_htlc` so thought it's okay to re-use the generic record?

File: lnwire/lnwire_test.go [hunk 14]
```diff
@@ -1337,6 +1337,42 @@ func TestLightningWireProtocol(t *testing.T) {
...
+				pubkey, err := randPubKey()
+				if err != nil {
+					t.Fatalf("unable to generate key: %v",
+						err)
+
+					return
+				}
+
+				req.BlindingPoint = tlv.SomeRecordT(
+					tlv.NewPrimitiveRecord[tlv.TlvType0](
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply lnwire/lnwire_test.go]: hm, strange this patch compiles for me on `go1.22.1 linux/amd64` [tlv0.patch.txt](https://github.com/lightningnetwork/lnd/files/14691594/tlv0.patch.txt)

File: lnwire/update_add_htlc.go [hunk 15]
```diff
@@ -78,6 +78,19 @@ type UpdateAddHTLC struct {
 	ExtraData ExtraOpaqueData
 }
 
+// BlindingPointOrNil returns the blinding point associated with the update, or
+// nil.
+func (c *UpdateAddHTLC) BlindingPointOrNil() *btcec.PublicKey {
```
Comment (Roasbeef) [inline lnwire/update_add_htlc.go:83]: So this can just be `c.BlindingPoint.UnwrapOr(nil)`.

Comment (carlaKC) [reply lnwire/update_add_htlc.go]: `cannot use nil as tlv.RecordT[*tlv.tlvType0, *secp256k1.PublicKey] value in argument to htlc.BlindingPoint.UnwrapOr`? 

iiuc we need two layers of unwrapping here to get the underlying type in the record rather than the type in the option?

Comment (Roasbeef) [reply lnwire/update_add_htlc.go]: Ah sure, had just glanced at that. Can also use `WhenSomeV` instead here to avoid needing to deref `Val`, but non-blocking ofc. 

Philosophical point, but generally I like to review the option as sort of lazy evaluation, so you pipe it down all the way until you actually need to obtain the true value. Otherwise, by forcing evaluation sooner than needed, you just end up burdening the caller once again with required nil checks.

File: peer/brontide.go [hunk 16]
```diff
@@ -1985,8 +1986,19 @@ func messageSummary(msg lnwire.Message) string {
 			msg.FeeSatoshis)
 
 	case *lnwire.UpdateAddHTLC:
-		return fmt.Sprintf("chan_id=%v, id=%v, amt=%v, expiry=%v, hash=%x",
-			msg.ChanID, msg.ID, msg.Amount, msg.Expiry, msg.PaymentHash[:])
+		var blindingPoint []byte
+		msg.BlindingPoint.WhenSome(
```
Comment (Roasbeef) [inline peer/brontide.go:1990]: Can be written as: 
```go
blindingPoint := fn.MapOptionz(msg.BlindingPoint, func(b ..) []byte {
    return b.Val.SerializeCompressed()
})
```

Non-blocking.

File: channeldb/channel.go [hunk 17]
```diff
@@ -2340,6 +2393,12 @@ func SerializeHtlcs(b io.Writer, htlcs ...HTLC) error {
 	}
 
 	for _, htlc := range htlcs {
+		// Populate TLV stream for any additional fields contained
+		// in the TLV.
+		if err := htlc.serializeExtraData(); err != nil {
```
Comment (Roasbeef) [inline channeldb/channel.go:2398]: Won't this information already be stored in `ExtraData` if you just take the opaque bytes from the update add HTLC message? Since we code, but then keep the blob as is.

File: lnwallet/channel.go [hunk 18]
```diff
@@ -736,6 +744,14 @@ func (c *commitment) toDiskCommit(ourCommit bool) *channeldb.ChannelCommitment {
 			Incoming:      false,
 		}
 		copy(h.OnionBlob[:], htlc.OnionBlob)
+		if htlc.BlindingPoint != nil {
```
Comment (Roasbeef) [inline lnwallet/channel.go:747]: Re above (and also related to future changes like storing the endorsement bit in an opaque manner), if we copied the extra records from the pay desc into this HTLC, then it's a more generic way to handle storing any future TLV data.

File: lnwallet/channel.go [hunk 19]
```diff
@@ -859,6 +882,12 @@ func (lc *LightningChannel) diskHtlcToPayDesc(feeRate chainfee.SatPerKWeight,
 		theirWitnessScript: theirWitnessScript,
 	}
 
+	htlc.BlindingPoint.WhenSome(func(b tlv.RecordT[
```
Comment (Roasbeef) [inline lnwallet/channel.go:885]: `WhenSomeV` is useful here again as it gives you just `*btcec.PublicKey` in this case.

File: lnwallet/channel.go [hunk 20]
```diff
@@ -3607,6 +3638,14 @@ func (lc *LightningChannel) createCommitDiff(
 				PaymentHash: pd.RHash,
 			}
 			copy(htlc.OnionBlob[:], pd.OnionBlob)
+			if pd.BlindingPoint != nil {
```
Comment (Roasbeef) [inline lnwallet/channel.go:3641]: Similar point here re just copying over the raw bytes so you don't need to be concerned about the record mapping at this state.

File: lnwallet/channel.go [hunk 21]
```diff
@@ -3736,12 +3775,21 @@ func (lc *LightningChannel) getUnsignedAckedUpdates() []channeldb.LogUpdate {
 		// four messages that it corresponds to.
 		switch pd.EntryType {
 		case Add:
+			var b lnwire.BlindingPointRecord
+			if pd.BlindingPoint != nil {
```
Comment (Roasbeef) [inline lnwallet/channel.go:3779]: If the pay desc just stores the record directly, then we also don't need to handle this shuffling.

File: contractcourt/htlc_incoming_contest_resolver.go [hunk 22]
```diff
@@ -520,9 +522,18 @@ func (h *htlcIncomingContestResolver) Supplement(htlc channeldb.HTLC) {
 func (h *htlcIncomingContestResolver) decodePayload() (*hop.Payload,
 	[]byte, error) {
 
+	var blindingPoint *btcec.PublicKey
+	h.htlc.BlindingPoint.WhenSome(
```
Comment (Roasbeef) [inline contractcourt/htlc_incoming_contest_resolver.go:526]: `WhenSomeV`

**Review by Roasbeef — APPROVED** [review]
LGTM 🦋

Nice work! 

Left some non-blocking comments re slightly different wayts to use some of the new APIs. The other thing that jumped out is that we can get away with just writing/reading `ExtraData` in a few more places to not have to worry about ensuring that all the new HTLC TLVs are properly written/read from disk. One example is when we add the endorsement bit directly or indirectly to `lnd`, the logic now also needs to make sure that's passed along vs just passing the `TLV` blob then leaving it to the caller at the last moment to unpack things. A similar comment applies to some of the option usage as well.

File: lnwallet/channel.go (hunk 18 above)
Comment (carlaKC) [reply lnwallet/channel.go:747]: > if we copied the extra records from the pay desc into this HTLC

def, but iirc from the original PR we wanted to be more intentional about what we store (ie, only things we care about) rather than including the full `ExtraBytes` and wasting space if people send us random junk along with the HTLC.
